File: session-02-dependency-injection/typescript/src/order/postgres-order.repository.ts
import {Injectable} from "@nestjs/common";
import {Order} from "./order.model";
import {OrderRepository} from "./order.repository";

// Exercise 3: Swap implementation. Change useClass in order.module.ts to use this.
@Injectable()
export class PostgresOrderRepository extends OrderRepository {
    private readonly orders: Order[] = [
        {id: 1, drink: "Flat White", status: "ready"},
        {id: 2, drink: "Mocha", status: "preparing"},
        {id: 3, drink: "Cold Brew", status: "pending"},
        {id: 4, drink: "Matcha Latte", status: "ready"},
    ];

    findAll(): Order[] {
        return this.orders;
    }

    findById(id: number): Order | undefined {
        return this.orders.find((o) => o.id === id);
    }
}

File: session-02-dependency-injection/java/src/main/java/coffeeshop/order/PostgresOrderRepository.java
package coffeeshop.order;

import org.springframework.context.annotation.Primary;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;

// Exercise 3: Swap the repository.
// @Primary makes this the default OrderRepository bean.
// Spring picks this over InMemoryOrderRepository without changing OrderService.
@Repository
@Primary
public class PostgresOrderRepository implements OrderRepository {

    private final List<Order> orders = List.of(
            new Order(1, "Flat White", "ready"),
            new Order(2, "Mocha", "preparing"),
            new Order(3, "Cold Brew", "pending"),
            new Order(4, "Matcha Latte", "ready")
    );

    @Override
    public List<Order> findAll() {
        return orders;
    }

    @Override
    public Optional<Order> findById(int id) {
        return orders.stream()
                .filter(order -> order.id() == id)
                .findFirst();
    }
}
